text/0000-proc-macro-include.md

```diff
@@ -0,0 +1,175 @@
+- Feature Name: `proc_macro_include`
+- Start Date: 2021-11-24
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Proc macros can now effectively `include!` other files and process their contents.
+This both allows proc macros to communicate that they read external files,
+and to maintain spans into the external file for more useful error messages.
+
+# Motivation
+[motivation]: #motivation
+
+- `include!` and `include_str!` are no longer required to be compiler built-ins,
+  and could be implemented as proc macros.
+- Help incremental builds and build determinism, by proc macros telling rustc which files they read.
+- Improve proc macro sandboxability and cacheability, by offering a way to implement this class of
+  file-reading macros without using OS APIs directly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## For users of proc macros
+
+Nothing changes! You'll just see nicer errors and fewer rebuilds
+from procedural macros which read external files.
+
+## For writers of proc macros
+
+Three new functions are provided in the `proc_macro` interface crate:
+
+```rust
+/// Read the contents of a file as a `TokenStream` and add it to build dependency info.
+///
+/// The build system executing the compiler will know that the file was accessed during compilation,
+/// and will be able to rerun the build when the contents of the file changes.
+///
+/// May fail for a number of reasons, for example, if the string contains unbalanced delimiters
+/// or characters not existing in the language.
+///
+/// If the file fails to be read, this is not automatically a fatal error. The proc macro may
+/// gracefully handle the missing file, or emit a compile error noting the missing dependency.
+///
+/// Source spans are constructed for the read file. If you use the spans of this token stream,
+/// any resulting errors will correctly point at the tokens in the read file.
+///
+/// NOTE: some errors may cause panics instead of returning `io::Error`.
+/// We reserve the right to change these errors into `io::Error`s later.
+fn include<P: AsRef<str>>(path: P) -> Result<TokenStream, std::io::Error>;
```

Comment:
There are three logical steps here - read the file into bytes (`std::fs::read`, can fail with `std::io::Error`), convert bytes into a string (`std::str::from_utf8`, can fail with `std::str::Utf8Error`), and parse string into a token stream (`proc_macro::TokenStream::from_str`, can fail with `proc_macro::LexError`).
It's somewhat strange to merge all of this into a single `std::io::Error`.

Reply:
`io::Error` does [support an arbitrary `impl Error + Send + Sync` payload](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each), so it's not completely that odd of a choice.

It might make more sense to give a custom error type, though, so we can potentially give it an `.emit()` method (or however the proc_macro diagnostics API ends up going) to emit a nice error.

I am kind of split with providing a simple-to-stabilize API, though.